resources/idlharness-shadowrealm.js

```diff
@@ -0,0 +1,78 @@
+ * running the code in a `ShadowRealm`, avoiding repetitive boilerplate.
+ *
+ * @see https://github.com/tc39/proposal-shadowrealm
+ * @param {String[]} srcs Spec name(s) for source idl files (fetched from
+ *      /interfaces/{name}.idl).
+ * @param {String[]} deps Spec name(s) for dependency idl files (fetched
+ *      from /interfaces/{name}.idl). Order is important - dependencies from
+ *      each source will only be included if they're already know to be a
+ *      dependency (i.e. have already been seen).
+ */
+function idl_test_shadowrealm(srcs, deps) {
+    const script_urls = [
+        "/resources/testharness.js",
+        "/resources/WebIDLParser.js",
+        "/resources/idlharness.js",
+    ];
+    promise_test(async t => {
+        const realm = new ShadowRealm();
+        realm.evaluate("globalThis.self = globalThis; undefined");
+
+        const ss = await Promise.all(script_urls.map(url => fetch_text(url)));
+        for (const s of ss) {
+            realm.evaluate(s);
+        }
+        const specs = await Promise.all(srcs.concat(deps).map(fetch_spec));
+        const idls = JSON.stringify(specs.map(i => i.idl));
+        const code = `
+            const idls = ${idls};
+            let results;
+            add_completion_callback(function (tests, harness_status, asserts_run) {
+                results = tests;
+            });
+
+            // Without the wrapping test, testharness.js will think it's done after it has run
+            // the first idlharness test.
+            test(() => {
+                const idl_array = new IdlArray();
+                for (let i = 0; i < ${srcs.length}; i++) {
+                    idl_array.add_idls(idls[i]);
+                }
+                for (let i = ${srcs.length}; i < ${srcs.length + deps.length}; i++) {
+                    idl_array.add_dependency_idls(idls[i]);
+                }
+                idl_array.test();
+            }, "inner setup");
+            String(JSON.stringify(results))
+        `;
+
+        // We ran the tests in the ShadowRealm and gathered the results. Now treat them as if
+        // we'd run them directly here, so we can see them.
+        const results = JSON.parse(realm.evaluate(code));
+        for (const {name, status, message} of results) {
+            // TODO: make this an API in testharness.js - needs RFC?
+            async_test(t => {t.set_status(status, message); t.phase = t.phases.HAS_RESULT; t.done()}, name);
+        }
+    }, "outer setup");
+}
+
+function fetch_text(url) {
+    return fetch(url).then(function (r) {
+        if (!r.ok) {
+            throw new IdlHarnessError("Error fetching " + url + ".");
+        }
+        return r.text();
+    });
+}
+
+/**
+ * fetch_spec is a shorthand for a Promise that fetches the spec's content.
+ */
+function fetch_spec(spec) {
+    var url = '/interfaces/' + spec + '.idl';
+    return fetch_text(url).then(idl => ({ spec, idl }));
```

Comment:
We don't actually use the `spec` property, so the `fetch_spec` wrapper is kind of silly. Can you just inline the URL construction into a `fetch_text` call instead?